145816283 application interface

thebere opened the pull request:
#### What does this PR do?
Completes task for building user interface

#### Description of Task to be completed?
1. Build interface for login/registration/logout
2. User been able to add documents
3. Application dashboard for managing documents
5. User page

#### How should this be manually tested?
N/A

#### Any background context you want to provide?
This task was completed using React, Material UI, and SASS.

#### What are the relevant pivotal tracker stories?
Build user interface (#145816283)

#### Screenshots (if appropriate)
###### Homepage
<img width="1279" alt="screen shot 2017-05-21 at 7 04 04 pm" src="https://cloud.githubusercontent.com/assets/26272640/26286270/f56f0b60-3e58-11e7-88d3-49595fe36f92.png">

###### Dashboard
<img width="1278" alt="screen shot 2017-05-21 at 7 04 33 pm" src="https://cloud.githubusercontent.com/assets/26272640/26286271/066fb02c-3e59-11e7-8e38-9c7f333792af.png">

###### Registration page
<img width="1280" alt="screen shot 2017-05-21 at 7 04 19 pm" src="https://cloud.githubusercontent.com/assets/26272640/26286275/11e36548-3e59-11e7-8308-6e4a417f1be0.png">

###### User profile page
<img width="1280" alt="screen shot 2017-05-21 at 7 04 56 pm" src="https://cloud.githubusercontent.com/assets/26272640/26286277/1ef9d9c4-3e59-11e7-9f3f-cc918f299cfb.png">

###### User update screen
<img width="1280" alt="screen shot 2017-05-21 at 7 05 07 pm" src="https://cloud.githubusercontent.com/assets/26272640/26286280/2cbe1958-3e59-11e7-80f3-9689ee7b5026.png">

###### View/search users screen
<img width="1273" alt="screen shot 2017-05-21 at 7 05 23 pm" src="https://cloud.githubusercontent.com/assets/26272640/26286283/38abd354-3e59-11e7-8d68-8456b83bef88.png">

###### Sidemenu / drawer Homepage
<img width="1279" alt="screen shot 2017-05-21 at 7 05 35 pm" src="https://cloud.githubusercontent.com/assets/26272640/26286289/45ef6c06-3e59-11e7-8b2e-485ddcdbbe14.png">



#### Questions:

N/A

Commit:
featute(routes): defining routes

[#145816283]

Commit:
feature(index.js): application root

[#145816283]

Commit:
feature(authentication): authenticating routes

using localstorage to persist data

[#145816283]

Commit:
feature(action): document actions

implement actions for document management

[#145816283]

Commit:
feature(action): search action

implement search action for searching users and documents

[#145816283]

Commit:
feature(action): signup action

implement signup action for registering users

[#145816283]

Commit:
feature(types): add action types

[#145816283]

Commit:
feature(action): users action

managing multiple users (profiles, update, etc)

[#145816283]

Commit:
feature(app root): app.jsx

defining application root

[#145816283]

Commit:
feature(documentStack): displaying documents

use to display multuple documents on a stack/grid format

[#145816283]

Commit:
feature(DocumentViewDialog): view document

view document in a dialog box

[#145816283]

Commit:
feature(edit dialog): edit dialog modal

[#145816283]

Commit:
feature(input text): define input fields

gloabal style for input

[#145816283]

Commit:
feature(documentModal): managinig document

modal for managing document action: edit and delete buttons and confirmation

[#145816283]

Commit:
feature(Dashboard): application dashboard

[#145816283]

Commit:
feature(dashboard component): infomation boxes

[#145816283]

Commit:
feature(private): private document component

[#145816283]

Commit:
feature(public): public document component
[#145816283]

Commit:
feature(public): public document component
[#145816283]

Commit:
feature(role): role document component

[#145816283]

Commit:
feeature(add document): add document page

[#145816283]

Commit:
feature(document form): add document form

[#145816283]

Commit:
feature(footer): application footer
[#145816283]

Commit:
feature(feed): home feed

feeding documents to the home page

[#145816283]

Commit:
featur(home.jsx): home page

with call to action buttons
[#145816283]

Commit:
feature(Index): index for components
[#145816283]

Commit:
feature(navigation): site navigation

top and side bars
[#145816283]

Commit:
feature(editprofile): edit profile form

[#145816283]

Commit:
feature(profile): user profile page

[#145816283]

Commit:
feature(profileInfo): user infomation component

[#145816283]

Commit:
feature(ProfileTabs): one page multiple tabs

use profile page

[#145816283]

Commit:
feature(UserDocuments): component for each user documents

[#145816283]

Commit:
feature(searchbox): top navigation box

[#145816283]

Commit:
feature(Signupform): signup form

[#145816283]

Commit:
feature(signin page): sign in page

[#145816283]

Commit:
feature(signup form): signup form

[#145816283]

Commit:
feature(signup page): signup conatiner page

[#145816283]

Commit:
feature(viewusers): table for viewing and searcing for users

[#145816283]

Commit:
feature(view users): view users container

[#145816283]

Commit:
chore(configs): babel, gitignore, webpack configs

[#145816283]

Commit:
feature(server): tryout static html

Commit:
feature(header token): fix axios header delete token

now user token gets deleted on signout

[#145816283]

Commit:
feature(HOC): controls routing

higher order component for controlliing access to specific routes

[#145816283]

Commit:
feature(HOC): higher order component
controlling routing access

[#145816283]

Commit:
feature(helpers): client helper methods

[#145816283]

Commit:
feature(store): configure redux store

[#145816283]

Commit:
feature(reducer): user reducer

[#145816283]

Commit:
feature(reducer): search reducer

[#145816283]

Commit:
feature(reducer): root reducer

[#145816283]

Commit:
feature(reducer): document reducer
[#145816283]

Commit:
feature(reducer): authentication reducer

[#145816283]

Commit:
feature(reducer): all users reducer

reducer handling multiple user actions

[#145816283]

Commit:
Merge branch 'develop' of https://github.com/Andela-JUdensi/DMS into test/145389353/integration-test

Commit:
chore(babelrc): remove state-0 support

[#145816283]

Commit:
documentation(readme.md): add documentation to readme

[#145823243]

femisenjobi commented on client/actions/documents.action.js:
```diff
@@ -0,0 +1,55 @@
+
+
+export const deleteDocumentAction = documentID => dispatch => axios.delete(`/api/documents/${documentID}`)
+  .then((response) => {
```
You may leave out the `response` since you're not using it. You can remove the `.catch` or handle the error.

femisenjobi commented on client/actions/documents.action.js:
```diff
@@ -0,0 +1,55 @@
+
+export const editDocumentAction = (documentID, documentData) => (dispatch) => {
+  return axios.put(`/api/documents/${documentID}`, documentData)
+    .then((response) => {
```
Same as above.

femisenjobi commented on client/actions/search.action.js:
```diff
@@ -0,0 +1,16 @@
+  .then((response) => {
+    dispatch(getSearchSuccess(response.data));
+  })
+  .catch((error) => {
```
Remove or handle the error

femisenjobi commented on client/components/App.jsx:
```diff
@@ -0,0 +1,26 @@
+import scss from '../assets/app.scss';
+import AppMenu from './menu/Index';
+
+class App extends React.Component {
```
This can be a stateless component since you do not need any life-cycle methods

thebere replied:
That's great. Will fix all that. And creating a container component for each, separating logic from presentation

femisenjobi commented on client/components/common/ViewDocumentInModal.jsx:
```diff
@@ -0,0 +1,164 @@
+  }
+
+  render() {
+    const { ownerID, access, id } = this.props.documentToView;
```
You can try out deep deconstructing like this:
```const { documentToView, documentToView: { ownerID, access, id, user: { documentOwnerRoleID } } } = this.props;```

thebere replied:
Oh. Never seen this one. But it worked like charm. Thanks

femisenjobi commented on client/components/common/ViewDocumentInModal.jsx:
```diff
@@ -0,0 +1,164 @@
+    const userLinks = (
+      <div>
+        {
+          (this.state.isAuthenticated)
```
To make the render method cleaner, this can live in a function outside the render method.

femisenjobi commented on client/components/common/ViewDocumentInModal.jsx:
```diff
@@ -0,0 +1,164 @@
+      </div>
+    );
+
+    const guestLinks = (
```
this can live outside the render method too in a `renderGuestLinks` method, for example, and call with `this.renderGuestLinks()`. Same with  `deleteDocementActions`.

femisenjobi commented on client/components/documents/AddDocument.jsx:
```diff
@@ -0,0 +1,20 @@
+import React from 'react';
+import DocumentForm from './DocumentForm';
+
+class AddDocument extends React.Component {
```
Feel free to switch this component to a stateless component

femisenjobi commented on client/components/home/Home.jsx:
```diff
@@ -0,0 +1,34 @@
+import { Link } from 'react-router-dom';
+import Feed from './Feed';
+
+class Home extends React.Component {
```
Stateless

femisenjobi commented on client/components/profile/EditProfileForm.jsx:
```diff
@@ -0,0 +1,213 @@
+      isLoading: false,
+      redirect: false,
+      snackBarOpen: false,
+      firstname: '',
```
Can you put all of the user state in to the user object
```
user: {}
```
and access it with `this.state.user[<property>]`

femisenjobi commented on client/components/profile/EditProfileForm.jsx:
```diff
@@ -0,0 +1,213 @@
+
+  componentDidMount() {
+    setTimeout(() => {
+      this.setState({
```
`this.setState({...this.props.user})` should work just fine. or `this.setState({user: {...this.props.user}})` if you think the chage above is okay.

femisenjobi commented:
Well done Joshua. It is obvious the amount of work you put in in such a short time. He is a summary of the review I made.
- Use stateless components when there is not need for state or life-cycle methods
- You may extract your styles into external files. Looks like you have that consistently in almost every component
- You can decongest most of your render methods. One way is to use renderFunctions that return the expected jsx
- Handle your errors or remove the catch blocks in your actions
- Fix your hound-ci errors

Try to implement the suggestions all across the code since almost the same things were repeated. Enjoy!

sayidu commented on README.md:
```diff
@@ -1,2 +1,380 @@
+  - [Delete document](#delete-document)
+- [Search](#search)
+  - [Search Documents](#search-documents)
+  - [Search Users] (#search-users)
```
Review this line.

sayidu commented on client/actions/authentication.action.js:
```diff
@@ -0,0 +1,45 @@
+    });
+};
+
+export const signInAction = userInput => dispatch => new Promise((resolve, reject) => {
```
Usage of promise is not required here, consider refactoring as such: 
```
export const signInAction = userInput => dispatch => axios.post('/api/users/login/', userInput)
    .then((response) => {
       const { token } = response.data;
       const { id, roleID } = response.data.user;
       localStorage.setItem('hermesToken', token);
       localStorage.setItem('hermesUserID', id);
       localStorage.setItem('hermesRoleID', roleID);
      setAuthorizationToken(token, id, roleID);
       dispatch(setCurrentUser(jwt.decode(token)));
      return response;
     })
     .catch(error => throw err);
```

sayidu commented on client/actions/documents.action.js:
```diff
@@ -0,0 +1,55 @@
+    dispatch(getDocumentSuccess(response.data));
+  })
+  .catch(() => {
+    // throw (error);
```
Take out all un-required commented lines.

thebere replied:
Thanks. done

sayidu commented on client/actions/users.action.js:
```diff
@@ -0,0 +1,70 @@
+    usersSearch,
+  };
+};
+export const searchUserAction = (query) => {
```
Inconsistent spacing.

Commit:
chores(fix chores): configuration

modify webpack production config

[#143922841]

Commit:
feature(index.html): add index

app index file for mounting react

[#145816283]

Commit:
feature(requireAuthentication): modify filename

[#145816283]

Commit:
feature(actions): implement redux actions

[#145816283]

Commit:
refactor(app.jsx): stateless

refactor component to a stateless component

[#145816283]

Commit:
refactor(documentOptions) change filename

refactor code to make readable

[#145816283]

Commit:
refactor(Feed.jsx): removes unncessary codes

[#145816283]

Commit:
refactor(styles): styles moves

moves styles to separate files and make stateless components
[#145816283]

Commit:
bug(fixed): determineDocsforUser

determineDocsforUser method to return correctly based on auth status and user priviledge

[#145816283]

Commit:
refactor(entry app): server and app
[#145816283]

Commit:
feature(reducers): implemets app reducers
[#145816283]

Commit:
feature(alerts.jsx): higer order component for showing alerts

[#145816283]

Commit:
feature(Helpers): client helper

add method for counting user document

[#145816283]

Commit:
feature(client auth): authenticate react/client routes

[#145816283]

Commit:
feature(routes): add middlewares to specific middleware routes

[#145816283]

Commit:
feature(types): define action types (redux)

[#145816283]

Commit:
bug(middleware): validate user inputs

[#145816283]

Commit:
refactor(documents contorller): recieve req data from middleware

[#145816283]

Commit:
styles(styles.js): add material ui styles

[#145816283]

Commit:
feature(routes): add middleware to routes

[#145816283]

Commit:
refactor(formTextInputes): removes styles

[#145816283]

Commit:
chore(chores): ignore folders with eslint

[#143921503]

Commit:
test(integration): refactor integration test

[#145389353]

Commit:
add engines

Commit:
deploy(procfile): add heroku procfile

[#146142105]

Commit:
deployment fixes: add post install

Commit:
feature(database.config): add production db url

[#146142105]

Commit:
deployment fixes: add post install

Commit:
feature(scss): add application style sheet

[#145816283]

Commit:
refactor(inde.html): remove duplicate link
 duplicate material icon links
[#145816283]

Commit:
deployment fixes: add post install

Commit:
refactor(dashboard): fix user document count

[#145816283]

Commit:
documentation(readme.md): add readme
[#145823243]

Commit:
bug(app.js): fix routing

[#146142105]

Commit:
refactor(app.js): comment out favicon error

[#146142105]

Commit:
test(packaage.json): fix errors from post install

[#143921565]

Commit:
bug(fixed)

Commit:
test(component): add component test
[#145389357]

Commit:
refactor(fix): fix text script
[#145389357]

Commit:
refactor(models): use set null rather than cascade

[#145389579]

Commit:
fix changes

Commit:
test(package.json): fix test script
